src/services/redis/backend.rs

```diff
@@ -0,0 +1,457 @@
+
+        // a reversed-level-order traversal delete
+        // recursion is not friendly for rust, and on-heap stack is preferred.
+        let (mut queue, mut rev_stack) = (VecDeque::new(), VecDeque::new());
+
+        queue.push_back(path.to_string());
+        while let Some(pop) = queue.pop_front() {
+            let skey = format!("v{}:k:{}", REDIS_API_VERSION, pop);
+            let mut it = con
+                .sscan(skey)
+                .await
+                .map_err(|err| new_exec_async_cmd_error(err, Operation::Delete, path.as_str()))?;
+            while let Some(child) = it.next_item().await {
+                let _ = queue.push_back(child);
+            }
+            let _ = rev_stack.push_back(pop);
+        }
+
+        while let Some(pop) = rev_stack.pop_back() {
+            remove(&mut con, abs_path.as_str(), pop.as_str()).await?;
+        }
+
+        if let Some(parent) = PathBuf::from(abs_path.clone()).parent() {
+            let p = parent.display().to_string();
+            let skey = format!("v{}:k:{}", REDIS_API_VERSION, p);
+            con.srem(skey, abs_path.clone()).await.map_err(|err| {
+                new_exec_async_cmd_error(err, Operation::Delete, abs_path.as_str())
+            })?;
+        }
+
+        Ok(())
+    }
+
+    async fn list(&self, path: &str, _args: OpList) -> Result<DirStreamer> {
+        let path = build_abs_path(self.root.as_str(), path);
+        let con = self
+            .client
+            .get_tokio_connection_manager()
+            .await
+            .map_err(|err| new_exec_async_cmd_error(err, Operation::List, path.as_str()))?;
+        Ok(Box::new(DirStream::new(
+            Arc::new(self.clone()),
+            con,
+            path.as_str(),
+        )))
+    }
+}
+
+fn build_abs_content_path<'a>(root: &'a str, rel_path: &'a str) -> String {
+    format!(
+        "v{}:c:{}",
+        REDIS_API_VERSION,
+        build_abs_path(root, rel_path)
+    )
+}
+
+/// remove all data of the `to_remove` entry
+async fn remove(con: &mut ConnectionManager, path: &str, to_remove: &str) -> Result<()> {
+    let m_path = format!("v{}:m:{}", REDIS_API_VERSION, to_remove);
+    let c_path = format!("v{}:c:{}", REDIS_API_VERSION, to_remove);
+    let s_path = format!("v{}:k:{}", REDIS_API_VERSION, to_remove);
+    con.del(&[m_path, c_path, s_path])
+        .await
+        .map_err(|err| new_exec_async_cmd_error(err, Operation::Delete, path))
+}
+
+fn build_abs_meta_path<'a>(root: &'a str, rel_path: &'a str) -> String {
+    format!(
+        "v{}:m:{}",
+        REDIS_API_VERSION,
+        build_abs_path(root, rel_path)
+    )
+}
+
+fn build_abs_to_child_path(root: &str, rel_path: &str) -> String {
+    format!(
+        "v{}:ch:{}",
+        REDIS_API_VERSION,
+        build_abs_path(root, rel_path)
+    )
+}
+
+#[cfg(test)]
+mod bin_test {
+    use crate::ObjectMetadata;
+
+    #[test]
+    fn test_generate_bincode() {
```

Comment:
Bad test case. Please assert with real data.

Reply:
The testing module is still incomplete, it's used to generate test data.

I'll remove the test flag in following pushes.